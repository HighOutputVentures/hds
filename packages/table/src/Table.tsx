import {
  Box,
  Center,
  chakra,
  Flex,
  HStack,
  Icon,
  Spinner,
  SystemStyleObject,
  Table,
  TableContainer,
  Tbody,
  Td,
  Text,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react';
import { Checkbox } from '@highoutput/hds-forms';
import { Tooltip } from '@highoutput/hds-tooltip';
import * as React from 'react';
import { v4 as uuid } from 'uuid';
import { useStyles } from './hooks';
import ArrowDownIcon from './icons/ArrowDownIcon';
import HelpCircleIcon from './icons/HelpCircleIcon';

export type UnknownArray = unknown[];
export type ArrayItem<T extends UnknownArray> = T[number];

export type SortDirection = 'asc' | 'desc';

export type SortContext = {
  direction: SortDirection;
};

export type CheckContext<T> = {
  item: T;
  isChecked: boolean;
};

export type ClickContext<T> = {
  item: T;
};

export type Column<T extends UnknownArray> = {
  label: string;
  width?: string;
  tooltip?: React.ReactNode;
  renderRow?: (item: T[number]) => any;
  onSort?(context: SortContext): void;
  onClick?(context: ClickContext<ArrayItem<T>>): void;
  onCheck?(context: CheckContext<ArrayItem<T>>): void;
  onCheckAll?(context: CheckAllContext<T>): void;
  defaultSort?: SortDirection;
  defaultChecked?: ((item: ArrayItem<T>) => boolean) | boolean;
  isHidden?: boolean;
  /**
   *
   * NOT IMPLEMENTED
   *
   */
  isSticky?: boolean;
};

export type CheckAllContext<T extends UnknownArray> = {
  selected: ArrayItem<T>[];
  isChecked: boolean;
};

export type TableBaseProps<T extends UnknownArray> = {
  items: T;
  columns: Column<T>[];
  isLoading?: boolean;
  renderLoader?: JSX.Element;
  renderHeader?: React.ReactNode;
  renderFooter?: React.ReactNode;
  isBordered?: boolean;
};

export type TableProps<T extends UnknownArray> = TableBaseProps<T> &
  Omit<SystemStyleObject, Required<keyof TableBaseProps<T>>>;

export default function HdsTable<T extends UnknownArray>(props: TableProps<T>) {
  const {
    items,
    columns,
    isLoading,
    renderLoader,
    renderHeader,
    renderFooter,
    isBordered,
    borderColor = 'Gray.200',
    ...others
  } = props;

  const styles = useStyles({
    isBordered,
    borderColor,
    hasBottomRowBorder: !!renderFooter,
  });

  const [checkedItems, setCheckedItems] = React.useState(() => {
    return columns.map(({ defaultChecked = false }) => {
      return items.map((item) => {
        const fn =
          typeof defaultChecked === 'boolean' ? () => defaultChecked : defaultChecked;
        return fn(item);
      });
    });
  });

  const totalColumns = columns.length;
  const shouldShowTable = items.length >= 1;
  const shouldShowEmpty = items.length <= 0 && !isLoading;
  const shouldHardLoad = items.length <= 0 && !!isLoading;
  const shouldSoftLoad = items.length >= 1 && !!isLoading;

  const SoftLoader = () => (!renderLoader ? <SoftLoaderDefault /> : renderLoader);
  const HardLoader = () =>
    !renderLoader ? <HardLoaderDefault numOfCols={totalColumns} /> : renderLoader;

  return (
    <Box
      sx={{
        ...others,
        border: '1px solid',
        borderColor,
        rounded: 'md',
        position: 'relative',
      }}
    >
      {!!renderHeader && (
        <Box
          sx={{
            borderBottom: '1px solid',
            borderColor,
            paddingX: '24px',
            paddingY: '20px',
          }}
        >
          {renderHeader}
        </Box>
      )}

      {shouldSoftLoad && <SoftLoader />}

      <TableContainer
        data-testid="hds.table.container"
        aria-busy={isLoading}
        sx={styles.container}
      >
        <Table
          data-testid="hds.table"
          suppressHydrationWarning
          sx={styles.table}
          position={'sticky'}
          left={0}
          right={0}
        >
          <Thead data-testid="hds.table.header">
            <Tr data-testid="hds.table.header.tr">
              {columns
                .filter((col) => !col.isHidden)
                .map(
                  (
                    {
                      //
                      label,
                      tooltip,
                      width,
                      onSort,
                      onCheck,
                      isSticky,
                      onCheckAll,
                      defaultSort = 'desc',
                    },
                    index,
                  ) => {
                    const isCheckedAll = !!checkedItems.at(index)?.every((o) => !!o);
                    const isIndeterminate =
                      !isCheckedAll && !!checkedItems.at(index)?.some((o) => !!o);

                    return (
                      <Th
                        key={uuid()}
                        width={width}
                        position={'sticky'}
                        {...(index !== 0 &&
                          isSticky && {
                            right: 0,
                          })}
                        {...(index === 0 &&
                          isSticky && {
                            left: 0,
                          })}
                        zIndex={isSticky ? columns.length - index : 0}
                        data-testid="hds.table.header.th"
                        bgColor={isSticky ? '#F9FAFB' : 'unset'}
                        filter={'drop-shadow(-4px 0px 8px rgba(0, 0, 0, 0.04))'}
                      >
                        <Flex alignItems="center">
                          {!!onCheck && (
                            <Checkbox
                              aria-label="Select all"
                              marginRight="12px"
                              size="sm"
                              isChecked={isCheckedAll}
                              isIndeterminate={isIndeterminate}
                              onChange={(e) => {
                                const isChecked = e.target.checked;

                                const selected = checkedItems[index].reduce<
                                  ArrayItem<T>[]
                                >((arr, bool, index_1) => {
                                  /*
                                   * false will still be set to true,
                                   * so we need to reverse the logic
                                   */
                                  return !bool ? [...arr, items[index_1]] : arr;
                                }, []);

                                onCheckAll?.({ isChecked, selected });
                                setCheckedItems((i) => {
                                  return i.map((j, i) => {
                                    return i !== index
                                      ? j
                                      : j.map(() => {
                                          return isChecked;
                                        });
                                  });
                                });
                              }}
                            />
                          )}
                          {label}
                          {onSort && <SortButton value={defaultSort} onSort={onSort} />}
                          {tooltip && (
                            <Tooltip label={tooltip} hasArrow>
                              <chakra.div display="flex" ml="4px">
                                <Icon
                                  as={HelpCircleIcon}
                                  color="neutrals.500"
                                  width="16px"
                                  height="16px"
                                />
                              </chakra.div>
                            </Tooltip>
                          )}
                        </Flex>
                      </Th>
                    );
                  },
                )}
            </Tr>
          </Thead>

          <Tbody position="relative" data-testid="hds.table.body">
            {shouldShowTable &&
              items.map((item, index_0) => {
                return (
                  <Tr key={uuid()} data-testid="hds.table.body.tr">
                    {columns
                      .filter((col) => !col.isHidden)
                      .map(
                        (
                          {
                            onSort,
                            onCheck,
                            onClick,
                            isSticky,
                            defaultChecked,
                            ...others
                          },
                          index_1,
                        ) => {
                          const renderRow = others.renderRow ?? ((obj) => String(obj));

                          return (
                            <Td
                              key={uuid()}
                              onClick={() => {
                                onClick?.({ item });
                              }}
                              position={'sticky'}
                              {...(index_1 !== 0 &&
                                isSticky && {
                                  right: 0,
                                })}
                              {...(index_1 === 0 &&
                                isSticky && {
                                  left: 0,
                                })}
                              zIndex={isSticky ? columns.length - index_1 : 0}
                              bgColor={isSticky ? 'white' : 'unset'}
                              data-testid="hds.table.body.td"
                              filter={'drop-shadow(-4px 0px 8px rgba(0, 0, 0, 0.04))'}
                            >
                              <Flex alignItems="center" gap="12px">
                                {onCheck && (
                                  <Checkbox
                                    aria-label="Select item"
                                    size="sm"
                                    isChecked={checkedItems[index_1][index_0]}
                                    onChange={(e) => {
                                      const isChecked = e.target.checked;
                                      onCheck({ isChecked, item });
                                      setCheckedItems((i) => {
                                        return i.map((j, idx_0) => {
                                          return idx_0 === index_1
                                            ? j.map((k, idx_1) =>
                                                idx_1 === index_0 ? isChecked : k,
                                              )
                                            : j;
                                        });
                                      });
                                    }}
                                  />
                                )}

                                <Box>{renderRow(item)}</Box>
                              </Flex>
                            </Td>
                          );
                        },
                      )}
                  </Tr>
                );
              })}

            {shouldHardLoad && <HardLoader />}
            {shouldSoftLoad && <SoftLoaderOverlay />}
            {shouldShowEmpty && <Empty numOfCols={totalColumns} />}
          </Tbody>
        </Table>
      </TableContainer>

      {!!renderFooter && (
        <Box paddingX="24px" paddingY="12px">
          {renderFooter}
        </Box>
      )}
    </Box>
  );
}

function Empty({ numOfCols }: { numOfCols: number }) {
  return (
    <Tr data-testid="hds.table.empty-indicator">
      <Td colSpan={numOfCols}>
        <Center width="full">
          <Text size="sm">No records found</Text>
        </Center>
      </Td>
    </Tr>
  );
}

function SoftLoaderDefault() {
  return (
    <Spinner
      pos="absolute"
      size="lg"
      zIndex="3"
      top="50%"
      left="50%"
      transform="translate(-50%, -50%)"
      color="brand.primary.700"
      emptyColor="neutrals.200"
      thickness="3px"
      speed="0.75s"
      data-testid="hds.table.loading-indicator"
    />
  );
}

function SoftLoaderOverlay() {
  return (
    <Tr
      position="absolute"
      zIndex="2"
      bgColor="whiteAlpha.700"
      top={0}
      left={0}
      width="full"
      height="full"
    />
  );
}

function HardLoaderDefault({ numOfCols }: { numOfCols: number }) {
  return (
    <Tr data-testid="hds.table.loading-indicator">
      <Td colSpan={numOfCols}>
        <Center width="full">
          <HStack spacing={2}>
            <Spinner
              size="sm"
              color="brand.primary.700"
              emptyColor="neutrals.200"
              thickness="3px"
              speed="0.75s"
            />
            <Text size="sm" color="neutrals.500">
              Loading...
            </Text>
          </HStack>
        </Center>
      </Td>
    </Tr>
  );
}

function SortButton({
  value,
  onSort,
}: {
  /** default value */
  value: SortDirection;
  onSort(ctx: SortContext): void;
}) {
  const [currentSort, setCurrentSort] = React.useState(value);
  // next sort value
  const direction = currentSort === 'asc' ? 'desc' : 'asc';

  return (
    <chakra.button
      role="button"
      aria-label={`Sort ${direction}`}
      display="flex"
      marginLeft="4px"
      transition="transform 300ms ease-in-out"
      transform={currentSort === 'asc' ? 'rotate(180deg)' : undefined}
      onClick={() => {
        setCurrentSort(direction);
        onSort({ direction });
      }}
    >
      <Icon as={ArrowDownIcon} color="Gray.500" width="16px" height="16px" />
    </chakra.button>
  );
}
