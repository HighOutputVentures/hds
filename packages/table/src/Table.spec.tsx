import { Avatar, HStack, Text } from '@chakra-ui/react';
import { faker } from '@faker-js/faker';
import '@testing-library/jest-dom';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import * as React from 'react';
import Table from './Table';

const items = new Array(5).fill(null).map(() => {
  return {
    id: faker.datatype.uuid(),
    name: faker.name.fullName(),
    email: faker.internet.email(),
    avatar: faker.internet.avatar(),
  };
});

const onSort = jest.fn();
const onCheck = jest.fn();
const onClick = jest.fn();
const onCheckAll = jest.fn();

describe('Table', () => {
  afterEach(cleanup);
  beforeEach(() => {
    render(
      <Table
        items={items}
        columns={[
          {
            label: 'Name',
            renderRow({ name, avatar }) {
              return (
                <HStack spacing={2}>
                  <Avatar src={avatar} name={name} />
                  <Text>{name}</Text>
                </HStack>
              );
            },
            onCheck,
            onCheckAll,
            onClick,
          },
          {
            label: 'Email',
            renderRow({ email }) {
              return email;
            },
            onSort,
            onClick,
            defaultSort: 'asc',
            tooltip: 'Hint: This is your email',
          },
        ]}
      />,
    );
  });

  describe('Checkbox', () => {
    it('Should render master checkbox', () => {
      expect(
        screen.queryByRole('checkbox', {
          name: 'Select all',
        }),
      ).toBeDefined();
    });

    it('Should render checkboxes', () => {
      expect(
        screen.queryAllByRole('checkbox', {
          name: 'Select item',
        }),
      ).toHaveLength(items.length);
    });

    it('Should call onCheck callback when a checkbox is toggled', () => {
      fireEvent.click(
        screen.getAllByRole('checkbox', {
          name: 'Select item',
        })[0],
      );

      expect(onCheck).toHaveBeenCalledWith(
        expect.objectContaining({
          isChecked: expect.any(Boolean),
          item: expect.any(Object),
        }),
      );
    });

    it('Should call onCheckAll callback when master checkbox is toggled', () => {
      fireEvent.click(
        screen.getByRole('checkbox', {
          name: 'Select all',
        }),
      );

      expect(onCheckAll).toHaveBeenCalledWith(
        expect.objectContaining({
          isChecked: expect.any(Boolean),
          selected: expect.any(Array),
        }),
      );
    });
  });

  describe('Sort', () => {
    it('Should render sort button', () => {
      expect(screen.queryByRole('button', { name: /Sort (a|de)sc/i })).toBeDefined();
    });

    it('Should call sort callback when sort button is clicked', () => {
      fireEvent.click(
        screen.getByRole('button', {
          name: /Sort (a|de)sc/i,
        }),
      );

      expect(onSort).toHaveBeenCalled();
    });
  });

  describe('Tooltip', () => {
    it('Should render help icon', () => {
      expect(screen.queryByRole('button', { name: 'Hint' })).toBeDefined();
    });
  });

  it('Should call onClick callback when rows with handlers attached are clicked', () => {
    const td = document.querySelector('td') as HTMLTableCellElement;
    fireEvent.click(td);
    expect(onClick).toHaveBeenCalled();
  });
});
