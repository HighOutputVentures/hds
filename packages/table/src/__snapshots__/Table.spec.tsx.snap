// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Should match snapshot 1`] = `
<div
  className="css-1uzs8zz"
>
  <div
    className="chakra-table__container css-noq6c0"
    data-testid="hds.table.container"
  >
    <table
      className="chakra-table css-1lravjh"
      data-testid="hds.table"
    >
      <thead
        className="css-0"
        data-testid="hds.table.header"
      >
        <tr
          className="css-0"
          data-testid="hds.table.header.tr"
        >
          <th
            className="css-0"
            data-testid="hds.table.header.th"
          >
            <div
              className="css-70qvj9"
            >
              <label
                className="chakra-checkbox css-1cd8jem"
                onClick={[Function]}
              >
                <input
                  aria-label="Select all"
                  checked={false}
                  className="chakra-checkbox__input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    Object {
                      "border": "0",
                      "clip": "rect(0, 0, 0, 0)",
                      "height": "1px",
                      "margin": "-1px",
                      "overflow": "hidden",
                      "padding": "0",
                      "position": "absolute",
                      "whiteSpace": "nowrap",
                      "width": "1px",
                    }
                  }
                  type="checkbox"
                />
                <span
                  aria-hidden={true}
                  className="chakra-checkbox__control css-dnty2r"
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                />
              </label>
              Name
            </div>
          </th>
          <th
            className="css-0"
            data-testid="hds.table.header.th"
          >
            <div
              className="css-70qvj9"
            >
              Email
              <button
                aria-label="Sort desc"
                className="css-sl2x72"
                onClick={[Function]}
                role="button"
              >
                <svg
                  className="chakra-icon css-8jpexl"
                  fill="none"
                  focusable={false}
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7.99992 3.33325V12.6666M7.99992 12.6666L12.6666 7.99992M7.99992 12.6666L3.33325 7.99992"
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="1.33333"
                  />
                </svg>
              </button>
              <button
                aria-controls="popover-content-:r1:"
                aria-expanded={false}
                aria-haspopup="dialog"
                aria-label="Hint"
                className="css-1dqet0s"
                id="popover-trigger-:r1:"
                onClick={[Function]}
              >
                <svg
                  className="chakra-icon css-brlnl4"
                  fill="none"
                  focusable={false}
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    clipPath="url(#clip0_5775_9734)"
                  >
                    <path
                      d="M6.05992 5.99998C6.21665 5.55442 6.52602 5.17872 6.93322 4.9394C7.34042 4.70009 7.81918 4.61261 8.2847 4.69245C8.75022 4.7723 9.17246 5.01433 9.47664 5.37567C9.78081 5.737 9.94729 6.19433 9.94659 6.66665C9.94659 7.99998 7.94659 8.66665 7.94659 8.66665M7.99992 11.3333H8.00659M14.6666 7.99998C14.6666 11.6819 11.6818 14.6666 7.99992 14.6666C4.31802 14.6666 1.33325 11.6819 1.33325 7.99998C1.33325 4.31808 4.31802 1.33331 7.99992 1.33331C11.6818 1.33331 14.6666 4.31808 14.6666 7.99998Z"
                      stroke="currentColor"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="1.33333"
                    />
                  </g>
                  <defs>
                    <clipPath
                      id="clip0_5775_9734"
                    >
                      <rect
                        fill="white"
                        height="16"
                        width="16"
                      />
                    </clipPath>
                  </defs>
                </svg>
              </button>
              <div
                className="chakra-popover__popper css-0"
                style={
                  Object {
                    "inset": "0 auto auto 0",
                    "minWidth": "max-content",
                    "position": "absolute",
                    "visibility": "hidden",
                  }
                }
              >
                <section
                  className="chakra-popover__content css-1deuoi4"
                  id="popover-content-:r1:"
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "--popper-arrow-shadow-color": undefined,
                      "--popper-arrow-size": undefined,
                      "opacity": 0,
                      "transform": "scale(0.95) translateZ(0)",
                      "transformOrigin": "var(--popper-transform-origin)",
                      "visibility": "hidden",
                    }
                  }
                  tabIndex={-1}
                >
                  <div
                    className="chakra-popover__arrow-positioner css-0"
                    data-popper-arrow=""
                    style={
                      Object {
                        "position": "absolute",
                      }
                    }
                  >
                    <div
                      className="chakra-popover__arrow css-0"
                      data-popper-arrow-inner=""
                    />
                  </div>
                  <div
                    className="chakra-popover__body css-0"
                    id="popover-body-:r1:"
                  >
                    Hint: This is your email
                  </div>
                </section>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        className="css-79elbk"
        data-testid="hds.table.body"
      >
        <tr
          className="css-0"
          data-testid="hds.table.body.tr"
        >
          <td
            className="css-0"
            data-testid="hds.table.body.td"
            onClick={[Function]}
          >
            <div
              className="css-120nera"
            >
              <label
                className="chakra-checkbox css-1elxtiq"
                onClick={[Function]}
              >
                <input
                  aria-label="Select item"
                  checked={false}
                  className="chakra-checkbox__input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    Object {
                      "border": "0",
                      "clip": "rect(0, 0, 0, 0)",
                      "height": "1px",
                      "margin": "-1px",
                      "overflow": "hidden",
                      "padding": "0",
                      "position": "absolute",
                      "whiteSpace": "nowrap",
                      "width": "1px",
                    }
                  }
                  type="checkbox"
                />
                <span
                  aria-hidden={true}
                  className="chakra-checkbox__control css-dnty2r"
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                />
              </label>
              <div
                className="css-0"
              >
                <div
                  className="chakra-stack css-15lgeks"
                >
                  <span
                    className="chakra-avatar css-1ruhx4w"
                  >
                    <div
                      aria-label="John"
                      className="chakra-avatar__initials css-0"
                      role="img"
                    >
                      J
                    </div>
                  </span>
                  <p
                    className="chakra-text css-0"
                  >
                    John
                  </p>
                </div>
              </div>
            </div>
          </td>
          <td
            className="css-0"
            data-testid="hds.table.body.td"
            onClick={[Function]}
          >
            <div
              className="css-120nera"
            >
              <div
                className="css-0"
              >
                john@domain.blah
              </div>
            </div>
          </td>
        </tr>
        <tr
          className="css-0"
          data-testid="hds.table.body.tr"
        >
          <td
            className="css-0"
            data-testid="hds.table.body.td"
            onClick={[Function]}
          >
            <div
              className="css-120nera"
            >
              <label
                className="chakra-checkbox css-1elxtiq"
                onClick={[Function]}
              >
                <input
                  aria-label="Select item"
                  checked={false}
                  className="chakra-checkbox__input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    Object {
                      "border": "0",
                      "clip": "rect(0, 0, 0, 0)",
                      "height": "1px",
                      "margin": "-1px",
                      "overflow": "hidden",
                      "padding": "0",
                      "position": "absolute",
                      "whiteSpace": "nowrap",
                      "width": "1px",
                    }
                  }
                  type="checkbox"
                />
                <span
                  aria-hidden={true}
                  className="chakra-checkbox__control css-dnty2r"
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                />
              </label>
              <div
                className="css-0"
              >
                <div
                  className="chakra-stack css-15lgeks"
                >
                  <span
                    className="chakra-avatar css-1ruhx4w"
                  >
                    <div
                      aria-label="mary"
                      className="chakra-avatar__initials css-0"
                      role="img"
                    >
                      m
                    </div>
                  </span>
                  <p
                    className="chakra-text css-0"
                  >
                    mary
                  </p>
                </div>
              </div>
            </div>
          </td>
          <td
            className="css-0"
            data-testid="hds.table.body.td"
            onClick={[Function]}
          >
            <div
              className="css-120nera"
            >
              <div
                className="css-0"
              >
                mary@domain.blah
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
